### PR TITLE
First implementation of Sensors system

### PR DESCRIPTION
**Pull request by traversaro**
With @ommac we had several issue while debugging some function related to ft regressors, and we lost an entire day without reaching a conclusion. 

Most of the problems are related to the use of the `FTSensor` and `FTSensorList` classes, that were an ugly hack that I quickly implemented more than one year ago, and I never had the time to clean up. Thanks to the TGV, I was able to refactor some code and to introduce a proper class system for representing sensor information. At the moment this is simplifying the debugging of inertial parameters regressors , but in the long term it can be useful for the C++ implementation of Berdy (hence cc @iron76 @nunoguedelha @claudia-lat @jeljaik @naveenoid ). 

@francesco-romano can you quickly review the `SensorsTree`/`SensorsMeasurements`/.. system? My polymorphism is quite harsh and I abused some concepts here and there, but can you check that I didn't do anything of completely wrong?

### CODE REVIEW

File: core/include/kdl_codyco/sensors.hpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+#include <string>
+
+namespace KDL {
+namespace CoDyCo {
+
+    enum SensorType
+    {
+        SIX_AXIS_FORCE_TORQUE = 0,
+        NR_OF_SENSOR_TYPES = 1
```
Comment (francesco-romano) [inline core/include/kdl_codyco/sensors.hpp]: Just a personal taste: I don't like to define the possible number of enum values in the enum itself.

File: core/include/kdl_codyco/sensors.hpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+             * sensor are added to the sensorsTree.
+             *
+             * The passed sensor will be dynamic casted to the specified sensor type,
+             * and will be copied in the sensors tree only if the dynamic cast will be successful.
+             *
+             * @param[in] sensor_type the SensorType of the added sensor.
+             * @param[in] a constant reference to the Sensor to add.
+             * @return the sensor index of the newly added sensor, or -1 in case of error.
+             */
+            int addSensor(const SensorType & sensor_type, const Sensor & sensor);
```
Comment (francesco-romano) [inline core/include/kdl_codyco/sensors.hpp]: Usually if you have to explicitly specify a type in a polymorphic hierarchy you are doing something wrong.

Try to remain the more abstract you can.
For example in this case add a 
`virtual Sensor * clone();` method in the sensor class and use it to copy the object

Comment (traversaro) [inline core/include/kdl_codyco/sensors.hpp]: The problem is that for the current design the index is type specific (also for compatibility with the wbi): if you have two ft sensors and one accelerometer, the index for the two will be  0 1 for the two FT sensors  and 0 for the accelerometer.
Hence I save the objects in different arrays depending on the type. Do you think it make sense to drop the sensor type in the `addSensor` method and then query dynamically the type using the `getSensorType()` method or `typeinfo` stuff?

File: core/include/kdl_codyco/sensors.hpp (hunk 1 above)
Comment (traversaro) [inline core/include/kdl_codyco/sensors.hpp]: Ok, the alternative is an external constant? Until we have static_assert that could leads to inconsistency when you add a value to the enum.

File: core/include/kdl_codyco/sensors.hpp (hunk 2 above)
Comment (traversaro) [inline core/include/kdl_codyco/sensors.hpp]: Mhh.. using `getSensorType()` and clone as you suggested we can get rid of the `const SensorType & sensor_type` argument, at least in this method, thanks.

Comment (francesco-romano) [inline core/include/kdl_codyco/sensors.hpp]: At least in this way the interface is clean.  You can always change the implementation later. 
Regarding how you detect the type of sensor it's up to you. Maybe the fastest (I'm not a fan of C++ RTTI. It is too limited to be of some use).

File: core/include/kdl_codyco/sensors.hpp (hunk 1 above)
Comment (francesco-romano) [inline core/include/kdl_codyco/sensors.hpp]: In my opinion the inconsistency can also occur in this case (and with the same probability).
But as I said before: it's just a personal taste.

File: core/include/kdl_codyco/sensors.hpp (hunk 2 above)
Comment (traversaro) [inline core/include/kdl_codyco/sensors.hpp]: Done in https://github.com/robotology-playground/idyntree/commit/cbba345b64ba70fd7b9a45d9a96911da51ce58e5

File: core/include/kdl_codyco/sensors.hpp (hunk 1 above)
Comment (traversaro) [inline core/include/kdl_codyco/sensors.hpp]: Done in https://github.com/robotology-playground/idyntree/commit/cbba345b64ba70fd7b9a45d9a96911da51ce58e5 .

**Comment by traversaro** [comment]
Everything works fine on the real robot and the test are passing, so I merge to avoid balkanization of the branches. 

Comments and suggestions are obvious still more then welcome.

**Pull request merged**